### PR TITLE
Add filter on task status in addition to desired status (Docker Provider - swarm)

### LINKED ISSUE #1238: Degraded performance in swarm during a scale up
### What is your environment & configuration? 
Docker swarm mode.
Multiple traefik instances, multiple services.
Using traefik.backend.loadbalancer.swarm=false for a service (or just keeping it anything other then true).

### What did you do?
Scaling up a service.

### What did you expect to see?
Everything works and performs well.

### What did you see instead?
The service endpoint have significant degraded performance during the new containers startup.

This issue doesn't reproduce when using traefik.backend.loadbalancer.swarm=true.
I don't have much experience with Traefik or swarm, but it looks to me that swarm's tasks don't change their state to running (from starting) until health-check is complete.
While swarm VIP is aware of the difference between state "starting" and "running" and does not pass traffic to these containers, I think that Traefik as of 1.2.0-rc2 ignores it.

### PR DESCRIPTION
**Pull request by Yshayy**
This PR solves the issue of passing traffic to a container before it's ready when using docker provider with swarm-mode and Traefik load-balancer (traefik.backend.loadbalancer.swarm=false), which is currently (V1.2-rc2) the default. (See issue #1238)

Current Tasks filter is based only on checking desired state, this PR adds an additional check on the current task state. 
For reference, check lifecycle model in swarmkit: https://github.com/docker/swarmkit/blob/master/design/task_model.md#task-lifecycle

Fixes #1238

### CODE REVIEW

**Comment by emilevauge** [comment]
@Yshayy Thanks for this :)
Could we have a unit test ?

**Comment by Yshayy** [comment]
Sure.
I'm not familiar with the docker test-suite, but it seems simple enough :)

File: provider/docker.go [hunk 1]
```diff
@@ -688,6 +688,9 @@ func listTasks(ctx context.Context, dockerClient client.APIClient, serviceID str
 	var dockerDataList []dockerData
 
 	for _, task := range taskList {
+		if task.Status.State != swarm.TaskStateRunning {
+			continue
+		}
```
Comment (emilevauge) [inline provider/docker.go]: One quick question why is this needed if we already [filtered](https://github.com/containous/traefik/pull/1304/files#diff-0e8a9f9593585fb3fa7ef0396fd5bbecR682) by `running` state before ?

**Comment by emilevauge** [comment]
I added a unit test.
/cc @containous/traefik

File: provider/docker_test.go [hunk 2]
```diff
@@ -2271,3 +2275,93 @@ func TestSwarmTaskParsing(t *testing.T) {
 		}
 	}
 }
+
+type fakeTasksClient struct {
+	dockerClient.APIClient
+	tasks []swarm.Task
+	err   error
```
Comment (vdemeester) [inline provider/docker_test.go]: Looking at the test, you don't need that one 👼

Comment (emilevauge) [inline provider/docker_test.go]: oops, indeed ^^

Comment (emilevauge) [inline provider/docker_test.go]: In fact, as it is used in return statement of `TaskList`, we may leave it as is.

File: provider/docker_test.go [hunk 3]
```diff
@@ -6,11 +6,15 @@ import (
 	"testing"
 
 	"github.com/containous/traefik/types"
+	"github.com/davecgh/go-spew/spew"
+	dockerClient "github.com/docker/engine-api/client"
```
Comment (timoreimann) [inline provider/docker_test.go]: Should be `dockerclient`. Go eschews camel case or even underscores in package names/aliases.

File: provider/docker_test.go [hunk 4]
```diff
@@ -2271,3 +2275,93 @@ func TestSwarmTaskParsing(t *testing.T) {
...
+
+func (c *fakeTasksClient) TaskList(ctx context.Context, options dockertypes.TaskListOptions) ([]swarm.Task, error) {
+	return c.tasks, c.err
+}
+
+func TestListTasks(t *testing.T) {
+	cases := []struct {
+		service       swarm.Service
+		tasks         []swarm.Task
+		isGlobalSvc   bool
```
Comment (timoreimann) [inline provider/docker_test.go]: Should be `isGlobalSVC`.

File: provider/docker_test.go [hunk 5]
```diff
@@ -2271,3 +2275,93 @@ func TestSwarmTaskParsing(t *testing.T) {
...
+			networks: map[string]*docker.NetworkResource{
+				"1": {
+					Name: "foo",
+				},
+			},
+		},
+	}
+
+	for _, e := range cases {
+		dockerData := parseService(e.service, e.networks)
```
Comment (timoreimann) [inline provider/docker_test.go]: I'd really like us to use more [sub-tests](https://blog.golang.org/subtests). Benefits:

- We can run sub-tests individually (`go test -run <sub test specifier>`)
- We are encouraged to pass a meaningful context into the `t.Run()` method (and, at the same time, don't need to add it at every `t.{Log,Error,Fatal}*` call).
- We can parallelize tests easily.
- If we use `t.Fatal*`, we only fatal out of the current sub-test and not the entire one.

Setting up sub-tests are really easy: [Example](https://github.com/timoreimann/traefik/blob/aae80ab324974cdb58f4cbed103caa15b0b3ba5e/provider/kubernetes_test.go#L328-L331).

File: provider/docker_test.go [hunk 6]
```diff
@@ -2271,3 +2275,93 @@ func TestSwarmTaskParsing(t *testing.T) {
...
+		dockerClient := &fakeTasksClient{tasks: e.tasks}
+		taskDockerData, _ := listTasks(context.Background(), dockerClient, e.service.ID, dockerData, map[string]*docker.NetworkResource{}, e.isGlobalSvc)
+
+		if len(e.expectedTasks) != len(taskDockerData) {
+			t.Fatalf("expected tasks %v, got %v", spew.Sprint(e.expectedTasks), spew.Sprint(taskDockerData))
+		}
+
+		for i, taskID := range e.expectedTasks {
+			if taskDockerData[i].Name != taskID {
+				t.Fatalf("expect task id %v, got %v", taskID, taskDockerData[i].Name)
```
Comment (timoreimann) [inline provider/docker_test.go]: Could this possibly be `t.Errorf`?

File: provider/docker_test.go [hunk 7]
```diff
@@ -2271,3 +2275,93 @@ func TestSwarmTaskParsing(t *testing.T) {
...
+					Name: "foo",
+				},
+			},
+		},
+	}
+
+	for _, e := range cases {
+		dockerData := parseService(e.service, e.networks)
+		dockerClient := &fakeTasksClient{tasks: e.tasks}
+		taskDockerData, _ := listTasks(context.Background(), dockerClient, e.service.ID, dockerData, map[string]*docker.NetworkResource{}, e.isGlobalSvc)
```
Comment (timoreimann) [inline provider/docker_test.go]: We should not ignore the error but check it and `t.Fatal` out if it turns out to be non-nil.

File: provider/docker_test.go [hunk 8]
```diff
@@ -2271,3 +2275,93 @@ func TestSwarmTaskParsing(t *testing.T) {
...
+
+		if len(e.expectedTasks) != len(taskDockerData) {
+			t.Fatalf("expected tasks %v, got %v", spew.Sprint(e.expectedTasks), spew.Sprint(taskDockerData))
```
Comment (timoreimann) [inline provider/docker_test.go]: I'd prefer `spew.Sdump` because

>To dump a variable with full newlines, indentation, type, and pointer information use Dump [...].

`Sprint` is a compacted form.

File: provider/docker_test.go (hunk 6 above)
Comment (emilevauge) [inline provider/docker_test.go]: @timoreimann On this, I suggest to make a global PR as a lot of test should be changed.

Comment (timoreimann) [inline provider/docker_test.go]: I agree for pre-existing tests (although we probably can't do this practically in a single PR only), but disagree and new ones. IMHO, there's no reason to create new test code with improper `t.Error`/`t.Fatal` usage.

File: provider/docker_test.go (hunk 7 above)
Comment (timoreimann) [inline provider/docker_test.go]: This is the only thing I find import to change now; my other comments would be nice to have but not critical.

File: provider/docker_test.go (hunk 6 above)
Comment (emilevauge) [inline provider/docker_test.go]: @timoreimann replaced in the whole file.

File: provider/docker_test.go (hunk 5 above)
Comment (emilevauge) [inline provider/docker_test.go]: Added sub tests in the whole file.

File: provider/docker_test.go [hunk 9]
```diff
@@ -149,12 +157,15 @@ func TestDockerGetFrontendRule(t *testing.T) {
...
 	}
 
-	for _, e := range containers {
-		dockerData := parseContainer(e.container)
-		actual := provider.getFrontendRule(dockerData)
-		if actual != e.expected {
-			t.Fatalf("expected %q, got %q", e.expected, actual)
-		}
+	for containerID, e := range containers {
+		t.Run(strconv.Itoa(containerID), func(t *testing.T) {
```
Comment (timoreimann) [inline provider/docker_test.go]: You need to make an assignment like `e := e` before calling `t.Run()` in order to capture the range variable. Otherwise, your variable won't get bound to the correct instance.

It's basically about [this particular Go gotcha](https://github.com/golang/go/wiki/CommonMistakes#using-goroutines-on-loop-iterator-variables).

Comment (timoreimann) [inline provider/docker_test.go]: Similar with the other tests.

Comment (emilevauge) [inline provider/docker_test.go]: Indeed, I forgot this awesome feature in go...

Thanks :+1:

**Comment by emilevauge** [comment]
@timoreimann fixes pushed :)

File: provider/docker_test.go (hunk 9 above)
Comment (timoreimann) [inline provider/docker_test.go]: ROFLing so hard right now...

**Comment by Yshayy** [comment]
Thanks
